flow-report/util.ts

```diff
@@ -0,0 +1,75 @@
+/**
+ * @license Copyright 2021 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import {createContext} from 'preact';
+import {useContext, useEffect, useState} from 'preact/hooks';
+
+export const FlowResultContext = createContext<LH.FlowResult|undefined>(undefined);
+
+function getHashParam(param: string): string|null {
+  const params = new URLSearchParams(location.hash.replace('#', '?'));
+  return params.get(param);
+}
+
+export function classNames(...args: Array<string|undefined|Record<string, boolean>>): string {
+  const classes = [];
+  for (const arg of args) {
+    if (!arg) continue;
+
+    if (typeof arg === 'string') {
+      classes.push(arg);
+      continue;
+    }
+
+    const applicableClasses = Object.entries(arg)
+      .filter(([_, shouldApply]) => shouldApply)
+      .map(([className]) => className);
+    classes.push(...applicableClasses);
+  }
+
+  return classes.join(' ');
+}
+
+export function useFlowResult(): LH.FlowResult {
+  // Expect this to always be called within a valid context provider.
+  // Cast to LH.FlowResult to prevent extra type handling.
+  return useContext(FlowResultContext) as LH.FlowResult;
```

Comment:
@patrickhulce WDYT about throwing an error here? Error should only throw because of developer misuse.

Alternative: `__LIGHTHOUSE_FLOW_JSON__` is already a global, so we could just access that directly instead of using a context.

Reply:
Let's not rely on the global :)

I'm fine throwing an error in situations like this where it is truly fatal and will happen on *every* render :) My earlier comment is about situations where there exists *some* reasonable fallback (like ignoring a details item or defaulting to no selection).